Create user migration script

kashizui opened the pull request:
This script migrates users from the Django user models to the bundle service
database.
1. Load user data from the Django database. Users with is_active=0 are not picked up.
2. ~~Generate a new id for each user using the uuid scheme~~
3. Update existing rows in the bundle service users table, using the Django
   user models as a source of truth, leaving only the disk/time quota and usage
   columns intact.
4. Insert remaining users into the bundle service users table, filling in the
   configured default quota information.
5. Mark all users in the Django database as inactive, which means that they
   will not be picked up in future runs of the migration script.

By default, this script runs in dry-run mode, i.e. it prints verbose output
such as the generate SQL queries but does not make changes to the databases.
When you're ready to perform the migration, run with the '-f' flag.

@klopyrev @percyliang

kashizui committed:
Create user migration script

kashizui committed:
Update documention

kashizui committed:
Update docs again

kashizui committed:
Commit deactivation query

kashizui committed:
Fix printing

kashizui commented:
Tested on both MySQL and sqlite, and works as advertised. Script is essentially idempotent.

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,237 @@
+###############################################################
+
+# Generate new user ids using uuid
+for user in django_users:
```
Keep the user IDs exactly the same. This user ID is used by the bundle service for it's permissions. If you change it, then all users will lose access to all their data.

kashizui replied:
Good point.

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,237 @@
+        with model.engine.begin() as connection:
+            insert_query = cl_user.insert().\
+                values(user_id=bindparam('new_user_id'),
+                       time_quota=default_user_info['time_quota'],
```
I'm sorry, but isn't this not populating the email and other information?

kashizui replied:
The other fields are automatically "bound" by sqlalchemy. (It inspects the keys in the first dict passed to `execute`. And it asserts that all the following dicts have exactly the same keys.)

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,237 @@
+###############################################################
+# Deactivate users in django db
+###############################################################
+
```
Why are you doing this?

kashizui replied:
So that running the script twice won't copy over users twice.

kashizui replied:
The `is_active` field is not actually used anywhere in the Django app, so this shouldn't impact anything else.

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,237 @@
+print "Users to update:", ', '.join(list(to_update))
+print "Users to insert:", ', '.join(list(to_insert))
+
+to_update = [user for user in django_users
```
Do away with this whole to_update and to_insert business. Just load up the entire Django database in memory. Load up the entire Bundle database in memory. Create a list of dicts containing all the new data that should go into the Bundle databases. This structure should contain the data from both databases. Then, in 1 transaction delete all data from the bundles database and insert all the new data.

kashizui replied:
Deleting users from the bundles database will not work, since there are foreign key constraints (the same reason we can't change the user IDs as you mentioned). Unfortunately it has to be separated into the insert and update steps, but makes sense to put it into one transaction.

kashizui committed:
Put everything in one transaction and don't change user id

kashizui committed:
Add comment about b_user_id

kashizui commented:
@klopyrev is this ok to merge or do you prefer a cleaner procedure?

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,209 @@
+    if django_config['database'].get('HOST', None):
+        db_params['host'] = django_config['database']['HOST']
+    if django_config['database'].get('PORT', None):
+        db_params['port'] = django_config['database']['PORT']
```
Tried testing this on wstest. The deployment.config file is a bit weird. It specifies the port as a string. 

Here's the fix:

```
if django_config['database'].get('PORT', None):
    db_params['port'] = django_config['database']['PORT']
    if isinstance(db_params['port'], unicode):
        db_params['port'] = int(db_params['port'])
```

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,209 @@
+        cluser.organization_or_affiliation AS affiliation,
+        cluser.is_superuser,
+        cluser.is_active,
+        email.verified AS is_verified
```
Change this line to

IF(NOT ISNULL(email.verified), email.verified, 0) AS is_verified

klopyrev commented on scripts/migrate-users.py:
```diff
@@ -0,0 +1,209 @@
+    # Deactivate users in django db
+    ###############################################################
+
+    if to_insert or to_update:
```
Don't do this. That makes it harder to roll back. Just delete this whole section.

klopyrev commented:
LGTM. Please address last comments before submitting.

kashizui committed:
Address migrate users comments

 - Don't deactivate users
 - Handle null is_verified
 - Ensure integer port number

kashizui committed:
Use proper fix for non-integer port
